packages/use-subscription/src/useSubscription.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {useEffect, useState} from 'react';
+
+// Hook used for safely managing subscriptions in concurrent mode.
+//
+// In order to avoid removing and re-adding subscriptions each time this hook is called,
+// the parameters passed to this hook should be memoized in some way–
+// either by wrapping the entire params object with useMemo()
+// or by wrapping the individual callbacks with useCallback().
+export function useSubscription<Value>({
+  // (Synchronously) returns the current value of our subscription.
+  getCurrentValue,
+
+  // This function is passed an event handler to attach to the subscription.
+  // It should return an unsubscribe function that removes the handler.
+  subscribe,
+}: {|
+  getCurrentValue: () => Value,
+  subscribe: (callback: Function) => () => void,
+|}): Value {
+  // Read the current value from our subscription.
+  // When this value changes, we'll schedule an update with React.
+  // It's important to also store the hook params so that we can check for staleness.
+  // (See the comment in checkForUpdates() below for more info.)
+  const [state, setState] = useState({
+    getCurrentValue,
+    subscribe,
+    value: getCurrentValue(),
```

Comment:
should this use lazy initial state to avoid calling `getCurrentValue()` every render as mentioned in https://github.com/facebook/react/pull/15022#issuecomment-497040407?

Reply:
Good call.